kubernetes/heapster.go

```diff
@@ -114,221 +111,26 @@ func getHeapsterURL(clientset kubernetes.Interface, clusterHostURL string) (URL
 
 }
 
-//ensure Heapster is accessible or launch a copy
-func ensureValidHeapster(config KubeAgentConfig) (KubeAgentConfig, error) {
-
-	client := &config.HTTPClient
-	var err error
-
-	// if Heapster is found in the cluster, test to make sure we can connect to it
-	// or if not using node summaries and unable to connect or find heapster in the kube-system namespace
-	// launch an instance of heapster in the cloudability namespace. Make sure and close the returned response body.
-	if config.HeapsterURL != "" {
-		return validateHeapster(config, client, config.Namespace)
-	}
-	if !config.RetrieveNodeSummaries {
-		log.Printf("Could not find heapster running in the cluster, launching a copy in the %s namespace.", config.Namespace)
-
-		config.HeapsterURL, err = launchHeapster(config.Clientset, config.UseInClusterConfig, config.Namespace)
-
-		if err != nil {
-			return config, fmt.Errorf("Error launching heapster in the %s namespace: %+v", err, config.Namespace)
-		}
-		innerTest, _, err := util.TestHTTPConnection(
-			client, config.HeapsterURL, http.MethodGet, config.BearerToken, retryCount, true)
-		if innerTest || err == nil {
-			log.Printf("Connected to heapster at: %v", config.HeapsterURL)
-		}
-	}
-
-	return config, err
-}
-
-func validateHeapster(config KubeAgentConfig, client rest.HTTPClient, namespace string) (KubeAgentConfig, error) {
+func validateHeapster(config KubeAgentConfig, client rest.HTTPClient) error {
 	outerTest, body, err := util.TestHTTPConnection(
 		client, config.HeapsterURL, http.MethodGet, config.BearerToken, retryCount, true)
 
-	var me heapsterMetricExport
+	if err != nil {
+		return err
+	}
 
+	var me heapsterMetricExport
 	if err := json.Unmarshal(*body, &me); err != nil {
 		log.Printf("malformed response from heapster running at: %v", config.HeapsterURL)
-	}
-
-	if !outerTest {
-		log.Printf("Could not connect to heapster at: %v, launching a copy in the Cloudability namespace.",
+	} else if !outerTest {
+		return fmt.Errorf("no heapster found")
+	} else if len(me) < 10 {
+		log.Printf("Received empty or malformed response from heapster running at: %v",
 			config.HeapsterURL)
```

Comment:
should this be an error return?

Reply:
Depends. Do we want this to be a fatal error? If we can connect to heapster but the response is malformed, do we want to give up on connecting to heapster in the future?

Reply:
seems like it should. if we're expecting to pull from heapster, but can't prove we're getting good data, it should error. at least from this method. i could see the caller deciding to be ok with it and just log it, if we're also collecting summaries and just "best effort" for heapster.